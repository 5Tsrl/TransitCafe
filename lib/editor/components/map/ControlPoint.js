import throttle from 'lodash.throttle'
import React, { Component, PropTypes } from 'react'
import { Marker } from 'react-leaflet'
import { shallowEqual } from 'react-pure-render'

export default class ControlPoint extends Component {
  static propTypes = {
    activePattern: PropTypes.object,
    icon: PropTypes.object,
    index: PropTypes.number,
    handleControlPointDrag: PropTypes.func,
    handleControlPointDragEnd: PropTypes.func,
    next: PropTypes.object,
    permanent: PropTypes.bool,
    position: PropTypes.array,
    previous: PropTypes.object,
    removeControlPoint: PropTypes.func,
    updateActiveEntity: PropTypes.func,
    updateControlPoint: PropTypes.func,
    updatePatternCoordinates: PropTypes.func
  }
  state = {
    latlng: null
  }
  shouldComponentUpdate (nextProps) {
    // TODO: fix this hack that keeps unknown position change (perhaps react-leaflet is updating it) from triggering
    // a component update, which funks with the position of the marker
    return !shallowEqual(nextProps.controlPoint.snap, this.props.controlPoint.snap)
  }
  _onClick = (e) => {
    console.log('control point clicked', e)
    // only remove controlPoint if it's not based on pattern stop (i.e., has permanent prop)
    if (!this.props.permanent) {
      this.props.removeControlPoint(this.props.activePattern, this.props.index, this.props.previous, this.props.next)
    }
  }
  handleDrag = () => {
    const latlng = this.refs.marker.leafletElement.getLatLng()
    this.setState({latlng})
    this.props.handleControlPointDrag(latlng, this.props.previous, this.props.next, this.props.activePattern)
  }
  _onDragEnd = (e) => {
    this.setState({latlng: null})
    this.props.handleControlPointDragEnd(this.props.index, this.props.controlPoint, e, this.props.activePattern)
  }
  render () {
    if (this.props.index === 1) console.log(this.props)
    const { position, icon } = this.props

    // keep track of position in state because we need this to be cleared once the user has
    // stopped dragging the marker, at which point this.state.latlng will be null and the marker
    // will "snap" back to the polyline
    const {latlng} = this.state
    const markerPosition = latlng ? [latlng.lat, latlng.lng] : position
    return (
      <Marker
        position={markerPosition}
        icon={icon}
        ref='marker'
        zIndexOffset={1000}
        draggable
        onDrag={throttle(this.handleDrag, 500)}
        onDragEnd={this._onDragEnd}
        onClick={this._onClick}
        color='black'
      />
    )
  }
}
