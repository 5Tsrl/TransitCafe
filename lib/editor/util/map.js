import ll from '@conveyal/lonlat'
import fetch from 'isomorphic-fetch'
import lineDistance from 'turf-line-distance'
import lineSlice from 'turf-line-slice'
import lineSliceAlong from '@turf/line-slice-along'
import lineString from 'turf-linestring'
import point from 'turf-point'
import pointOnLine from 'turf-point-on-line'

import {getSegment} from '../../scenario-editor/utils/valhalla'
import {generateUID} from '../../common/util/util'
import {getConfigProperty} from '../../common/util/config'
import {reverseEsri as reverse} from '../../scenario-editor/utils/reverse'

export const MAP_LAYERS = [
  {
    name: 'Streets',
    id: 'mapbox.streets'
  },
  {
    name: 'Light',
    id: 'mapbox.light'
  },
  {
    name: 'Dark',
    id: 'mapbox.dark'
  },
  {
    name: 'Satellite',
    id: 'mapbox.streets-satellite'
  }
]

export function stopToStopTime (stop) {
  return {stopId: stop.id, defaultDwellTime: 0, defaultTravelTime: 0}
}

export function clickToLatLng (latlng) {
  const precision = 100000000 // eight decimal places is accurate up to 1.1 meters
  return {stop_lat: Math.round(latlng.lat * precision) / precision, stop_lon: Math.round(latlng.lng % 180 * precision) / precision}
}

export function newControlPoint (distance, point, permanent = false, props = {}) {
  return {
    distance,
    id: generateUID(),
    permanent,
    point,
    ...props
  }
}

export async function constructStop (latlng, feedId) {
  const stopLatLng = clickToLatLng(latlng)
  const result = await reverse(latlng)
  const stopId = generateUID()
  let stopName = `New Stop (${stopId})`
  if (result && result.address) {
    stopName = result.address.Address
  }
  return {
    stop_id: stopId,
    stop_name: stopName,
    feedId,
    ...stopLatLng
  }
}

export function constructPoint (latlng) {
  const coords = ll.toCoordinates(latlng)
  return {
    type: 'Feature',
    geometry: {
      type: 'Point',
      coordinates: coords
    }
  }
}

export async function route (from, to) {
  const url = `${getConfigProperty('application.r5')}/plan?fromLat=${from.lat}&fromLon=${from.lng}&toLat=${to.lat}&toLon=${to.lng}&mode=CAR&full=false`
  const response = await fetch(url)
  return await response.json()
}

export async function street (from, to) {
  let json
  try {
    json = await route(from, to)
  } catch (e) {
    return null
  }
  if (json.errors) {
    console.log('error getting r5 data', json)
    return null
  } else {
    return json
  }
}

export function getPatternEndPoint (pattern) {
  const coordinates = pattern.shape && pattern.shape.coordinates
  const patternStops = [...pattern.patternStops]
  let endPoint
  if (coordinates) {
    endPoint = ll.toLeaflet(coordinates[coordinates.length - 1])
  } else {
    endPoint = {lng: patternStops[0].stop_lon, lat: patternStops[0].stop_lat}
  }
  return endPoint
}

export function getFeedBounds (feedSource, pad = 0) {
  return feedSource && feedSource.latestValidation && feedSource.latestValidation.bounds
    ? [[feedSource.latestValidation.bounds.north + pad, feedSource.latestValidation.bounds.west - pad], [feedSource.latestValidation.bounds.south - pad, feedSource.latestValidation.bounds.east + pad]]
    : [[60, 60], [-60, -20]]
}

/**
 * Recalculate the shape coordinates of a pattern after an edit
 *
 * @param  {ControlPoint[]}  controlPoints
 * @param  {String}  editType either 'insert, delete, or update'
 * @param  {Number}  index  The control point index.
 *   For insert, the point will be inserted at this position
 *   For delete, this point is removed and coordinates are recalculated between the remaining points
 *   For update, the coordinates are recalculated using the adjacent control points
 * @param  {Boolean}  followStreets
 * @param  {Boolean}  newPoint  Must be sent for insert operations
 * @param  {Boolean} [defaultToStraightLine=true]
 * @return {Coordinate[]} The resulting coordinates of the linestring of the updated pattern shape
 */
export async function recalculatePatternCoordinates (
  controlPoints,
  editType,
  index,
  followStreets,
  newPoint,
  patternShape,
  defaultToStraightLine = true
) {
  // validate arguments
  if (['insert', 'update'].indexOf(editType) > -1 && !newPoint) {
    throw new Error('newPoint argument is required when editType is "insert" or "update"')
  }

  /**
   * Convenience function to return the coordinates from a lineSliceAlong operation
   *
   * @param  {Number} startDistance
   * @param  {Number} endDistance
   * @return {Coordinate[]}
   */
  function lineSliceAlongToCoords (startDistance, endDistance) {
    return lineSliceAlong(
      lineString(patternShape.coordinates),
      startDistance,
      endDistance
    ).geometry.coordinates
  }

  const previousControlPoint = controlPoints[index - 1]
  const currentControlPoint = controlPoints[index]
  const followingControlPoint = controlPoints[index + 1]
  const lastControlPointDistance = mToKm(
    controlPoints[controlPoints.length - 1].distance
  )

  const newPointCoordinates = newPoint ? ll.toCoordinates(newPoint) : null

  const originalCoords = patternShape.coordinates
  let previousCoords = []
  let followingCoords = []

  // create points to make valhalla routing request
  // also store previous and remaining coords to make new resulting coordinate array
  let points

  if (editType === 'insert') {
    if (index === 0) {
      // inserting a point before first controlPoint
      points = [
        newPointCoordinates,
        currentControlPoint.point.geometry.coordinates
      ]
      followingCoords = originalCoords
    } else if (index === controlPoints.length) {
      // inserting a point at end of controlPoints
      points = [
        previousControlPoint.point.geometry.coordinates,
        newPointCoordinates
      ]
      previousCoords = originalCoords
    } else {
      // inserting a point between other control points
      points = [
        previousControlPoint.point.geometry.coordinates,
        newPointCoordinates,
        currentControlPoint.point.geometry.coordinates
      ]
      previousCoords = lineSliceAlongToCoords(
        0,
        mToKm(previousControlPoint.distance)
      )
      followingCoords = lineSliceAlongToCoords(
        mToKm(currentControlPoint.distance),
        lastControlPointDistance
      )
    }
  } else if (editType === 'update') {
    points = []
    if (index > 0) {
      points.push(previousControlPoint.point.geometry.coordinates)
      previousCoords = lineSliceAlongToCoords(
        0,
        mToKm(previousControlPoint.distance)
      )
    }
    points.push(newPointCoordinates)
    if (index < controlPoints.length - 1) {
      points.push(followingControlPoint.point.geometry.coordinates)
      followingCoords = lineSliceAlongToCoords(
        mToKm(followingControlPoint.distance),
        lastControlPointDistance
      )
    }
  } else if (editType === 'delete') {
    if (index === 0) {
      // deleting first stop.  Return coordinates of remaining stops
      return lineSliceAlongToCoords(
        mToKm(controlPoints[2].distance),
        lastControlPointDistance
      )
    } else if (index === controlPoints.length - 1) {
      // deleting last stop.  Return coordinates of all preceeding stops
      return lineSliceAlongToCoords(
        0,
        mToKm(controlPoints[controlPoints.length - 3].distance)
      )
    } else {
      // deleting an in-between control point
      points = [
        previousControlPoint.point.geometry.coordinates,
        followingControlPoint.point.geometry.coordinates
      ]
      previousCoords = lineSliceAlongToCoords(
        0,
        mToKm(previousControlPoint.distance)
      )
      followingCoords = lineSliceAlongToCoords(
        mToKm(followingControlPoint.distance),
        lastControlPointDistance
      )
    }
  }

  // calculate new segment
  const newSegment = await getSegment(
    points,
    followStreets,
    defaultToStraightLine
  )

  // rebuild linestring
  return [...previousCoords, ...newSegment.coordinates, ...followingCoords]
}

function geojsonFromCoordinates (coordinates) {
  return coordinates && {
    type: 'Feature',
    geometry: {
      coordinates: coordinates,
      type: 'LineString'
    }
  }
}

export function getControlPointSnap (controlPointGeoJson, coordinates) {
  const geojson = geojsonFromCoordinates(coordinates)

  // snap control point to line
  const snap = pointOnLine(geojson, controlPointGeoJson)
  const lineSegment = lineSlice(point(geojson.geometry.coordinates[0]), snap, geojson)

  // measure line segment
  const distTraveled = lineDistance(lineSegment, 'meters')
  return { snap, distTraveled }
}

/**
 * Translate meters to kilometers
 *
 * @param  {Number} m number of meters
 * @return {Number}   number of kilometers
 */
function mToKm (m) {
  return m / 1000
}
