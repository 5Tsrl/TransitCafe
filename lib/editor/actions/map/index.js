import shp from 'shpjs'
import lineSlice from 'turf-line-slice'
import point from 'turf-point'
import lineDistance from 'turf-line-distance'
import ll from '@conveyal/lonlat'

import {updateActiveGtfsEntity, saveActiveGtfsEntity} from '../active'
import {getControlPointSnap, handlePatternEdit, newControlPoint} from '../../util/map'
import {polyline as getPolyline} from '../../../scenario-editor/utils/valhalla'

export function updateMapSetting (props) {
  return {
    type: 'UPDATE_MAP_SETTING',
    props
  }
}

export function addControlPoint (controlPoint, index) {
  return {
    type: 'ADD_CONTROL_POINT',
    controlPoint,
    index
  }
}

export function handleControlPointDragEnd (index, controlPoint, evt, pattern) {
  return function (dispatch, getState) {
    const coordinates = getState().editor.editSettings.patternCoordinates
    const { snap, distTraveled } = getControlPointSnap(evt, coordinates)
    dispatch(updateActiveGtfsEntity(pattern, 'trippattern', {shape: {type: 'LineString', coordinates}}))
    dispatch(updateControlPoint(index, snap, distTraveled))
  }
}

export function handleControlPointDrag (latlng, previous, next, pattern) {
  return function (dispatch, getState) {
    const { followStreets, patternCoordinates } = getState().editor.editSettings
    const defaultToStraightLine = false
    handlePatternEdit(latlng, previous, next, pattern, followStreets, patternCoordinates, defaultToStraightLine)
    .then(coords => {
      if (coords) {
        dispatch(updatePatternCoordinates(coords))
      }
    })
  }
}

export function removingControlPoint (pattern, index, begin, end) {
  return {
    type: 'REMOVE_CONTROL_POINT',
    pattern,
    index,
    begin,
    end
  }
}

export function constructControlPoint (pattern, latlng, controlPoints) {
  return function (dispatch, getState) {
    // slice line
    const beginPoint = point(pattern.shape.coordinates[0])
    const coord = ll.toCoordinates(latlng)
    const clickPoint = point(coord)
    const lineSegment = lineSlice(beginPoint, clickPoint, pattern.shape)

    // measure line segment
    const distTraveled = lineDistance(lineSegment, 'meters')
    const controlPoint = newControlPoint(distTraveled, clickPoint)

    // find splice index based on shape dist traveled
    let index = 0
    for (var i = 0; i < controlPoints.length; i++) {
      if (distTraveled > controlPoints[i].distance) {
        index = i + 1
      } else {
        break
      }
    }
    // add control point
    dispatch(addControlPoint(controlPoint, index))
  }
}

export function updatePatternCoordinates (coordinates) {
  return {
    type: 'UPDATE_PATTERN_COORDINATES',
    coordinates
  }
}

export function removeControlPoint (pattern, index, begin, end) {
  return async function (dispatch, getState) {
    const { followStreets, patternCoordinates } = getState().editor.editSettings
    const coordinates = await handlePatternEdit(null, begin, end, pattern, followStreets, patternCoordinates)
    // update pattern
    dispatch(updateActiveGtfsEntity(pattern, 'trippattern', {shape: {type: 'LineString', coordinates}}))
    // remove controlPoint
    dispatch(removingControlPoint(pattern, index, begin, end))
  }
}

export function updateControlPoint (index, point, distance) {
  return {
    type: 'UPDATE_CONTROL_POINT',
    index,
    point,
    distance
  }
}

function receivedRoutesShapefile (feedSource, geojson) {
  return {
    type: 'RECEIVED_ROUTES_SHAPEFILE',
    feedSource,
    geojson
  }
}

export function displayRoutesShapefile (feedSource, file) {
  return function (dispatch, getState) {
    const reader = new window.FileReader()
    reader.onload = e => {
      const arrayBuff = reader.result
      shp(arrayBuff).then(geojson => {
        console.log(geojson)
        geojson.key = Math.random()
        dispatch(receivedRoutesShapefile(feedSource, geojson))
      })
    }
    reader.readAsArrayBuffer(file)
  }
}

export function extendPatternToPoint (pattern, endPoint, newEndPoint) {
  return async function (dispatch, getState) {
    let newShape = await getPolyline([endPoint, newEndPoint])

    // get single coordinate if polyline fails
    if (!newShape) {
      newShape = [ll.toCoordinates(newEndPoint)]
    }
    const updatedShape = {type: 'LineString', coordinates: [...pattern.shape.coordinates, ...newShape]}
    dispatch(updateActiveGtfsEntity(pattern, 'trippattern', {shape: updatedShape}))
    await dispatch(saveActiveGtfsEntity('trippattern'))
    return updatedShape
  }
}
