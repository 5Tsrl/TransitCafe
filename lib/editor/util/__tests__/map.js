/* globals describe, expect, it */

import nock from 'nock'

import {recalculatePatternCoordinates} from '../map'

const controlPoints = require('./test-control-points.json')
const followStreets = true
const patternShape = require('./test-pattern-shape.json')

describe('editor > util > map >', () => {
  describe('recalculatePatternCoordinates >', () => {
    function makeTestCase ({editType, index, mapzenResponeFile, newPoint}) {
      return async () => {
        // setup nock
        if (mapzenResponeFile) {
          nock('https://valhalla.mapzen.com')
            .get(/^\/route\?json.*/)
            .reply(200, require(mapzenResponeFile))
        }

        const coords = await recalculatePatternCoordinates(
          controlPoints,
          editType,
          index,
          followStreets,
          newPoint,
          patternShape
        )

        expect(coords.length).toBeGreaterThan(0)
        expect(coords).toMatchSnapshot()
      }
    }

    describe('inserting >', () => {
      it(
        'should insert a point before the first stop',
        makeTestCase({
          editType: 'insert',
          index: 0,
          mapzenResponeFile: './mapzen-response-1.json',
          newPoint: [-77.92553, 42.08666]
        })
      )

      it(
        'should insert a point after the last stop',
        makeTestCase({
          editType: 'insert',
          index: 5,
          mapzenResponeFile: './mapzen-response-2.json',
          newPoint: [-77.92262, 42.09828]
        })
      )

      it(
        'should insert a point in the middle of the control points',
        makeTestCase({
          editType: 'insert',
          index: 3,
          mapzenResponeFile: './mapzen-response-3.json',
          newPoint: [-77.92382, 42.09182]
        })
      )
    })

    describe('updating >', () => {
      it(
        'should update the first stop',
        makeTestCase({
          editType: 'update',
          index: 0,
          mapzenResponeFile: './mapzen-response-4.json',
          newPoint: [-77.92553, 42.08666]
        })
      )

      it(
        'should update the last stop',
        makeTestCase({
          editType: 'update',
          index: 4,
          mapzenResponeFile: './mapzen-response-5.json',
          newPoint: [-77.92262, 42.09828]
        })
      )

      it(
        'should update a point in the middle of the control points',
        makeTestCase({
          editType: 'update',
          index: 3,
          mapzenResponeFile: './mapzen-response-6.json',
          newPoint: [-77.92382, 42.09182]
        })
      )
    })

    describe('deleting >', () => {
      it(
        'should delete the first stop',
        makeTestCase({
          editType: 'delete',
          index: 0
        })
      )

      it(
        'should delete the last stop',
        makeTestCase({
          editType: 'delete',
          index: 4
        })
      )

      it(
        'should delete a point in the middle of the control points',
        makeTestCase({
          editType: 'delete',
          index: 3,
          mapzenResponeFile: './mapzen-response-7.json'
        })
      )
    })
  })
})
