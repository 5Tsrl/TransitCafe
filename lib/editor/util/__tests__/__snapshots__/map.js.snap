// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editor > util > map > recalculatePatternCoordinates > deleting > should delete a point in the middle of the control points 1`] = `
Array [
  Array [
    -77.922203,
    42.086407,
  ],
  Array [
    -77.922196,
    42.086498,
  ],
  Array [
    -77.922173,
    42.086498,
  ],
  Array [
    -77.92183700000001,
    42.086513000000004,
  ],
  Array [
    -77.92188300000001,
    42.086956,
  ],
  Array [
    -77.921929,
    42.087368,
  ],
  Array [
    -77.922043,
    42.088386,
  ],
  Array [
    -77.922112,
    42.088844,
  ],
  Array [
    -77.922135,
    42.089099,
  ],
  Array [
    -77.92231799999999,
    42.090522,
  ],
  Array [
    -77.92235600000001,
    42.090847000000004,
  ],
  Array [
    -77.922432,
    42.091491000000005,
  ],
  Array [
    -77.922509,
    42.092201,
  ],
  Array [
    -77.92252574804179,
    42.09235127309673,
  ],
  Array [
    -77.92252574804179,
    42.09235127309673,
  ],
  Array [
    -77.922524,
    42.092349999999996,
  ],
  Array [
    -77.922654,
    42.093502,
  ],
  Array [
    -77.922684,
    42.093745999999996,
  ],
  Array [
    -77.922707,
    42.09391,
  ],
  Array [
    -77.92274499999999,
    42.094158,
  ],
  Array [
    -77.922799,
    42.094528000000004,
  ],
  Array [
    -77.922974,
    42.095657,
  ],
  Array [
    -77.923058,
    42.096168,
  ],
  Array [
    -77.92305462145765,
    42.096168018547736,
  ],
  Array [
    -77.92305462145765,
    42.096168018547736,
  ],
]
`;

exports[`editor > util > map > recalculatePatternCoordinates > deleting > should delete the first stop 1`] = `
Array [
  Array [
    -77.92252574804179,
    42.09235127309673,
  ],
  Array [
    -77.922654,
    42.093502,
  ],
  Array [
    -77.922684,
    42.093745999999996,
  ],
  Array [
    -77.922707,
    42.09391,
  ],
  Array [
    -77.92274499999999,
    42.094158,
  ],
  Array [
    -77.922799,
    42.094528000000004,
  ],
  Array [
    -77.922974,
    42.095657,
  ],
  Array [
    -77.92305462145765,
    42.096168018547736,
  ],
]
`;

exports[`editor > util > map > recalculatePatternCoordinates > deleting > should delete the last stop 1`] = `
Array [
  Array [
    -77.922203,
    42.086407,
  ],
  Array [
    -77.922196,
    42.086498,
  ],
  Array [
    -77.922173,
    42.086498,
  ],
  Array [
    -77.92183700000001,
    42.086513000000004,
  ],
  Array [
    -77.92188300000001,
    42.086956,
  ],
  Array [
    -77.921929,
    42.087368,
  ],
  Array [
    -77.922043,
    42.088386,
  ],
  Array [
    -77.922112,
    42.088844,
  ],
  Array [
    -77.922135,
    42.089099,
  ],
  Array [
    -77.92231799999999,
    42.090522,
  ],
  Array [
    -77.92235600000001,
    42.090847000000004,
  ],
  Array [
    -77.922432,
    42.091491000000005,
  ],
  Array [
    -77.922509,
    42.092201,
  ],
  Array [
    -77.92252574804179,
    42.09235127309673,
  ],
]
`;

exports[`editor > util > map > recalculatePatternCoordinates > inserting > should insert a point after the last stop 1`] = `
Array [
  Array [
    -77.922203,
    42.086407,
  ],
  Array [
    -77.922196,
    42.086498,
  ],
  Array [
    -77.922173,
    42.086498,
  ],
  Array [
    -77.92183700000001,
    42.086513000000004,
  ],
  Array [
    -77.92188300000001,
    42.086956,
  ],
  Array [
    -77.921929,
    42.087368,
  ],
  Array [
    -77.922043,
    42.088386,
  ],
  Array [
    -77.922112,
    42.088844,
  ],
  Array [
    -77.922135,
    42.089099,
  ],
  Array [
    -77.92231799999999,
    42.090522,
  ],
  Array [
    -77.92235600000001,
    42.090847000000004,
  ],
  Array [
    -77.922432,
    42.091491000000005,
  ],
  Array [
    -77.922509,
    42.092201,
  ],
  Array [
    -77.922654,
    42.093502,
  ],
  Array [
    -77.922684,
    42.093745999999996,
  ],
  Array [
    -77.922707,
    42.09391,
  ],
  Array [
    -77.92274499999999,
    42.094158,
  ],
  Array [
    -77.922799,
    42.094528000000004,
  ],
  Array [
    -77.922974,
    42.095657,
  ],
  Array [
    -77.923104,
    42.096481,
  ],
  Array [
    -77.92305462145765,
    42.096168018547736,
  ],
  Array [
    -77.923058,
    42.096168,
  ],
  Array [
    -77.923104,
    42.096481,
  ],
  Array [
    -77.923226,
    42.096987999999996,
  ],
  Array [
    -77.92334,
    42.097350999999996,
  ],
  Array [
    -77.923577,
    42.097873,
  ],
  Array [
    -77.92363800000001,
    42.097988,
  ],
]
`;

exports[`editor > util > map > recalculatePatternCoordinates > inserting > should insert a point before the first stop 1`] = `
Array [
  Array [
    -77.92553,
    42.08666,
  ],
  Array [
    -77.92553000000001,
    42.086708,
  ],
  Array [
    -77.924515,
    42.086704,
  ],
  Array [
    -77.92405,
    42.086696,
  ],
  Array [
    -77.92369099999999,
    42.086704,
  ],
  Array [
    -77.923424,
    42.086696,
  ],
  Array [
    -77.923317,
    42.086687999999995,
  ],
  Array [
    -77.92319499999999,
    42.086662000000004,
  ],
  Array [
    -77.922921,
    42.086543999999996,
  ],
  Array [
    -77.922852,
    42.086528,
  ],
  Array [
    -77.92274499999999,
    42.086517,
  ],
  Array [
    -77.922196,
    42.086498,
  ],
  Array [
    -77.922203,
    42.086407,
  ],
  Array [
    -77.922196,
    42.086498,
  ],
  Array [
    -77.922173,
    42.086498,
  ],
  Array [
    -77.92183700000001,
    42.086513000000004,
  ],
  Array [
    -77.92188300000001,
    42.086956,
  ],
  Array [
    -77.921929,
    42.087368,
  ],
  Array [
    -77.922043,
    42.088386,
  ],
  Array [
    -77.922112,
    42.088844,
  ],
  Array [
    -77.922135,
    42.089099,
  ],
  Array [
    -77.92231799999999,
    42.090522,
  ],
  Array [
    -77.92235600000001,
    42.090847000000004,
  ],
  Array [
    -77.922432,
    42.091491000000005,
  ],
  Array [
    -77.922509,
    42.092201,
  ],
  Array [
    -77.922654,
    42.093502,
  ],
  Array [
    -77.922684,
    42.093745999999996,
  ],
  Array [
    -77.922707,
    42.09391,
  ],
  Array [
    -77.92274499999999,
    42.094158,
  ],
  Array [
    -77.922799,
    42.094528000000004,
  ],
  Array [
    -77.922974,
    42.095657,
  ],
  Array [
    -77.923104,
    42.096481,
  ],
]
`;

exports[`editor > util > map > recalculatePatternCoordinates > inserting > should insert a point in the middle of the control points 1`] = `
Array [
  Array [
    -77.922203,
    42.086407,
  ],
  Array [
    -77.922196,
    42.086498,
  ],
  Array [
    -77.922173,
    42.086498,
  ],
  Array [
    -77.92183700000001,
    42.086513000000004,
  ],
  Array [
    -77.92188300000001,
    42.086956,
  ],
  Array [
    -77.921929,
    42.087368,
  ],
  Array [
    -77.922043,
    42.088386,
  ],
  Array [
    -77.922112,
    42.088844,
  ],
  Array [
    -77.922135,
    42.089099,
  ],
  Array [
    -77.92231799999999,
    42.090522,
  ],
  Array [
    -77.92235600000001,
    42.090847000000004,
  ],
  Array [
    -77.922432,
    42.091491000000005,
  ],
  Array [
    -77.922509,
    42.092201,
  ],
  Array [
    -77.92252574804179,
    42.09235127309673,
  ],
  Array [
    -77.92252574804179,
    42.09235127309673,
  ],
  Array [
    -77.922524,
    42.092349999999996,
  ],
  Array [
    -77.922509,
    42.092201,
  ],
  Array [
    -77.92263799999999,
    42.092189000000005,
  ],
  Array [
    -77.923737,
    42.092182,
  ],
  Array [
    -77.92375200000001,
    42.091819,
  ],
  Array [
    -77.92375200000001,
    42.091819,
  ],
  Array [
    -77.923737,
    42.092182,
  ],
  Array [
    -77.92263799999999,
    42.092189000000005,
  ],
  Array [
    -77.922509,
    42.092201,
  ],
  Array [
    -77.922654,
    42.093502,
  ],
  Array [
    -77.922684,
    42.093745999999996,
  ],
  Array [
    -77.922707,
    42.09391,
  ],
  Array [
    -77.92274499999999,
    42.094158,
  ],
  Array [
    -77.92276100000001,
    42.094261,
  ],
  Array [
    -77.92276015989871,
    42.09426187381858,
  ],
  Array [
    -77.922799,
    42.094528000000004,
  ],
  Array [
    -77.922974,
    42.095657,
  ],
  Array [
    -77.92305462145765,
    42.096168018547736,
  ],
]
`;

exports[`editor > util > map > recalculatePatternCoordinates > updating > should update a point in the middle of the control points 1`] = `
Array [
  Array [
    -77.922203,
    42.086407,
  ],
  Array [
    -77.922196,
    42.086498,
  ],
  Array [
    -77.922173,
    42.086498,
  ],
  Array [
    -77.92183700000001,
    42.086513000000004,
  ],
  Array [
    -77.92188300000001,
    42.086956,
  ],
  Array [
    -77.921929,
    42.087368,
  ],
  Array [
    -77.922043,
    42.088386,
  ],
  Array [
    -77.922112,
    42.088844,
  ],
  Array [
    -77.922135,
    42.089099,
  ],
  Array [
    -77.92231799999999,
    42.090522,
  ],
  Array [
    -77.92235600000001,
    42.090847000000004,
  ],
  Array [
    -77.922432,
    42.091491000000005,
  ],
  Array [
    -77.922509,
    42.092201,
  ],
  Array [
    -77.92252574804179,
    42.09235127309673,
  ],
  Array [
    -77.92252574804179,
    42.09235127309673,
  ],
  Array [
    -77.922524,
    42.092349999999996,
  ],
  Array [
    -77.922509,
    42.092201,
  ],
  Array [
    -77.92263799999999,
    42.092189000000005,
  ],
  Array [
    -77.923737,
    42.092182,
  ],
  Array [
    -77.92375200000001,
    42.091819,
  ],
  Array [
    -77.92375200000001,
    42.091819,
  ],
  Array [
    -77.923737,
    42.092182,
  ],
  Array [
    -77.92263799999999,
    42.092189000000005,
  ],
  Array [
    -77.922509,
    42.092201,
  ],
  Array [
    -77.922654,
    42.093502,
  ],
  Array [
    -77.922684,
    42.093745999999996,
  ],
  Array [
    -77.922707,
    42.09391,
  ],
  Array [
    -77.92274499999999,
    42.094158,
  ],
  Array [
    -77.922799,
    42.094528000000004,
  ],
  Array [
    -77.922974,
    42.095657,
  ],
  Array [
    -77.923058,
    42.096168,
  ],
  Array [
    -77.92305462145765,
    42.096168018547736,
  ],
  Array [
    -77.92305462145765,
    42.096168018547736,
  ],
]
`;

exports[`editor > util > map > recalculatePatternCoordinates > updating > should update the first stop 1`] = `
Array [
  Array [
    -77.92553,
    42.08666,
  ],
  Array [
    -77.92553000000001,
    42.086708,
  ],
  Array [
    -77.924515,
    42.086704,
  ],
  Array [
    -77.92405,
    42.086696,
  ],
  Array [
    -77.92369099999999,
    42.086704,
  ],
  Array [
    -77.923424,
    42.086696,
  ],
  Array [
    -77.923317,
    42.086687999999995,
  ],
  Array [
    -77.92319499999999,
    42.086662000000004,
  ],
  Array [
    -77.922921,
    42.086543999999996,
  ],
  Array [
    -77.922852,
    42.086528,
  ],
  Array [
    -77.92274499999999,
    42.086517,
  ],
  Array [
    -77.922173,
    42.086498,
  ],
  Array [
    -77.92183700000001,
    42.086513000000004,
  ],
  Array [
    -77.92188300000001,
    42.086956,
  ],
  Array [
    -77.921929,
    42.087368,
  ],
  Array [
    -77.922043,
    42.088386,
  ],
  Array [
    -77.922112,
    42.088844,
  ],
  Array [
    -77.922135,
    42.089099,
  ],
  Array [
    -77.922158,
    42.089252,
  ],
  Array [
    -77.92215494185281,
    42.08925407008939,
  ],
  Array [
    -77.92231799999999,
    42.090522,
  ],
  Array [
    -77.92235600000001,
    42.090847000000004,
  ],
  Array [
    -77.922432,
    42.091491000000005,
  ],
  Array [
    -77.922509,
    42.092201,
  ],
  Array [
    -77.922654,
    42.093502,
  ],
  Array [
    -77.922684,
    42.093745999999996,
  ],
  Array [
    -77.922707,
    42.09391,
  ],
  Array [
    -77.92274499999999,
    42.094158,
  ],
  Array [
    -77.922799,
    42.094528000000004,
  ],
  Array [
    -77.922974,
    42.095657,
  ],
  Array [
    -77.92305462145765,
    42.096168018547736,
  ],
]
`;

exports[`editor > util > map > recalculatePatternCoordinates > updating > should update the last stop 1`] = `
Array [
  Array [
    -77.922203,
    42.086407,
  ],
  Array [
    -77.922196,
    42.086498,
  ],
  Array [
    -77.922173,
    42.086498,
  ],
  Array [
    -77.92183700000001,
    42.086513000000004,
  ],
  Array [
    -77.92188300000001,
    42.086956,
  ],
  Array [
    -77.921929,
    42.087368,
  ],
  Array [
    -77.922043,
    42.088386,
  ],
  Array [
    -77.922112,
    42.088844,
  ],
  Array [
    -77.922135,
    42.089099,
  ],
  Array [
    -77.92231799999999,
    42.090522,
  ],
  Array [
    -77.92235600000001,
    42.090847000000004,
  ],
  Array [
    -77.922432,
    42.091491000000005,
  ],
  Array [
    -77.922509,
    42.092201,
  ],
  Array [
    -77.922654,
    42.093502,
  ],
  Array [
    -77.922684,
    42.093745999999996,
  ],
  Array [
    -77.922707,
    42.09391,
  ],
  Array [
    -77.92274499999999,
    42.094158,
  ],
  Array [
    -77.92276015989871,
    42.09426187381858,
  ],
  Array [
    -77.92276015989871,
    42.09426187381858,
  ],
  Array [
    -77.923058,
    42.096168,
  ],
  Array [
    -77.923104,
    42.096481,
  ],
  Array [
    -77.923226,
    42.096987999999996,
  ],
  Array [
    -77.92334,
    42.097350999999996,
  ],
  Array [
    -77.923577,
    42.097873,
  ],
  Array [
    -77.92363800000001,
    42.097988,
  ],
]
`;
